### PR TITLE
Add Digit DP Module

### PR DESCRIPTION
**Pull request by jessechoe10**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: content/4_Gold/DP_Digits.mdx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	for (int i = 0; i < 19; i++) {
+		for (int j = 0; j < 50; j++) {
+			for (int k = 0; k < 2; k++) {
+				for (int l = 0; l < 2; l++) { dp[i][j][k][l] = -1; }
+			}
+		}
+	}
+}
+
+ll solveDP(int loc, int k, bool largest, bool started, int targ, int targ2) {
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Digits.mdx:88]: snake_case for c++

File: content/4_Gold/Digit_DP.problems.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        "kind": "USACO",
+        "usacoId": "1307"
+      }
+    },
+    {
+      "uniqueId": "usaco-1115",
+      "name": "Count the Cows",
+      "url": "http://usaco.org/index.php?page=viewproblem2&cpid=1115",
+      "source": "Gold",
+      "difficulty": "Very Hard",
```
Comment (bqi343) [inline content/4_Gold/Digit_DP.problems.json:37]: I don't get why this is "Very Hard" and the first problem is "Easy" - these problems aren't that different in difficulty IMO. maybe "Hard" for this one and "Normal" for the first one?

File: content/4_Gold/Digit_DP.mdx [hunk 3]
```diff
@@ -0,0 +1,149 @@
+---
+id: digit-dp
+title: 'Digit DP'
+author: Jesse Choe
+prerequisites:
+  - intro-dp
+  - modular
```
Comment (bqi343) [inline content/4_Gold/Digit_DP.mdx:7]: why are modular, combo, knapsack prereqs?

Comment (jessechoe10) [reply content/4_Gold/Digit_DP.mdx:7]: hmm ill just say intro dp then

**Review by bqi343 — APPROVED** [review]
looks okay (but I didn't review the explanation)

File: content/4_Gold/Digit_DP.mdx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+### Without Dynamic Programming
+
+We can solve this problem in $\mathcal{O}(Y - X)$ time by looping through $X \ldots Y$ and checking whether a given number is an interesting number. However, since $Y - X$ can be large, the naive solution is too slow, requiring us to optimize it.
+
+### With Dynamic Programming
+
+One way to optimize the brute force approach of looping from $X$ to $Y$ and counting interesting numbers is by using dynamic programming. The DP approach involves considering all 9 digits one at a time to occupy at least half of the number.
+
+Let $\texttt{dp[i][und][k][is0]}$ where $i$ is the current position, $und$ is whether you have gone below the actual number, $k$ is a counter to see if you have at least half of the same digit, and $is0$ is a boolean if you still have leading zeros. The transition involves looping through 9 digits to add to the current location and comparing with the digit we want to occupy at least half of the number.
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:56]: give a walkthrough of what a transition would look like or smth

File: content/4_Gold/Digit_DP.mdx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+This algorithm is fast enough as the number of digits is small and there are only nine digits.
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+// Code:
+// https://raw.githubusercontent.com/thecodingwizard/competitive-programming/master/USACO/2014open/silver/odometer.cpp
+
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:66]:
```suggestion
<CPPSection>

Code from
[here](https://raw.githubusercontent.com/thecodingwizard/competitive-programming/master/USACO/2014open/silver/odometer.cpp)

```cpp

File: content/4_Gold/Digit_DP.mdx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+
+ll dp[19][50][2][2];
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:72]: magic numbers

File: content/4_Gold/Digit_DP.mdx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		if (isStarted && targ2 != -1) {
+			if (i != targ && i != targ2) continue;
+		}
+		ans += solve_dp(loc + 1, k + (isStarted ? (targ == i ? 1 : -1) : 0),
+		                isLargest, isStarted, targ, targ2);
+	}
+	return dp[loc][k][largest][started] = ans;
+}
+
+ll solve(ll x) {
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:117]: `solve` gives me no info

File: content/4_Gold/Digit_DP.mdx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	num = to_string(x);
+	ll ans = 0;
+	for (int i = 0; i <= 9; i++) {
+		reset();
+		ans += solve_dp(0, 20, true, false, i, -1);
+	}
+	ll dup = 0;
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:124]: add some comments around here

File: content/4_Gold/Digit_DP.mdx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	if (dp[loc][k][largest][started] != -1) {
+		return dp[loc][k][largest][started];
+	}
+	ll ans = 0;
+	for (int i = 0; i <= 9; i++) {
+		bool isLargest = largest;
+		int lar = num[loc] - '0';
+		if (isLargest && i > lar) break;
+		if (i < lar) isLargest = false;
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:106]: braces + snake_case

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Digit_DP.mdx [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+Let $\texttt{dp[i][und][k][is0]}$ where $i$ is the current position, $und$ is whether you have gone below the actual number, $k$ is a counter to see if you have at least half of the same digit, and $is0$ is a boolean if you still have leading zeros. The transition involves looping through 9 digits to add to the current location and comparing with the digit we want to occupy at least half of the number.
+
+This algorithm is fast enough as the number of digits is small and there are only nine digits.
+
+<LanguageSection>
+<CPPSection>
+
+Code from
+[here](https://raw.githubusercontent.com/thecodingwizard/competitive-programming/master/USACO/2014open/silver/odometer.cpp).
```
Comment (ryanchou-dev) [inline content/4_Gold/Digit_DP.mdx:64]: better as "Code by Nathan Wang", and add link to github instead of just the raw file

**Review by ryanchou-dev — APPROVED** [review]
looks nice, just do sans' suggestions, and we'll be chillin

File: content/4_Gold/Digit_DP.mdx (hunk 9 above)
Comment (ryanchou-dev) [reply content/4_Gold/Digit_DP.mdx]: should be done ?

File: content/4_Gold/Digit_DP.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply content/4_Gold/Digit_DP.mdx:125]: get this
how about instead of adding a comment, you put the description in the name itself

**Comment by SansPapyrus683** [comment]
there's still unresolved comments my man

**Pull request closed without merging**